### ISSUE TITLE
Generator max length exeeded

### ISSUE THREAD

**Issue by Calypso971** [savonet/liquidsoap#3528]
Used Liquidsoap version: 2.2.2 on Debian Bookworm (12) as provided on Github.

My production machine is still running on 2.1.4, but I'm testing out v2.2.2 and experience some strange behaviour. The problem is that I can't reproduce it, but within the last couple of days it happened a few times: after connecting/disconnecting a live source Liquidsoap starts to consume lots of CPU resources. Most of the time it doesn't happen, but when (ofcourse) you least expect it, it's there again. I've tried with a small script, and I succeeded in triggering it once, however, I don't have a log of it.

The script that I've used and where it also happened is just 2 sources (1 local playlist, the other an input.harbor), a transition and a metadata rewrite for the live source.

I have a log of this morning with the normal script:

At first everything seem to be normal:

2023/11/13 09:54:54 [harbor:4] Header: Connection, value: Keep-Alive.
2023/11/13 09:54:54 [harbor:4] Header: Accept-Encoding, value: gzip, deflate.
2023/11/13 09:54:54 [harbor:4] Header: Accept-Language, value: en-NL,*.
2023/11/13 09:54:54 [harbor:4] Header: User-Agent, value: Mozilla/5.0.
2023/11/13 09:54:54 [harbor:4] http/unix GET request on /admin/metadata.
2023/11/13 09:54:54 [harbor:4] http/unix Arg: mount, value: /live.
2023/11/13 09:54:54 [harbor:4] http/unix Arg: charset, value: utf-8.
2023/11/13 09:54:54 [harbor:4] http/unix Arg: song, value: Just a test
2023/11/13 09:54:54 [harbor:4] http/unix Arg: mode, value: updinfo.
2023/11/13 09:54:54 [harbor:4] Request to update metadata for mount /live on port 8300
2023/11/13 09:54:54 [process:4] Starting process
2023/11/13 09:54:54 [process:4] Closing process's stdin
2023/11/13 09:54:54 [process:4] Process exited with code 0
2023/11/13 09:54:54 [process:4] Starting process
2023/11/13 09:54:54 [process:4] Cleaning up process
2023/11/13 09:54:54 [process:4] Closing process's stdin
2023/11/13 09:54:54 [process:4] Process exited with code 0
2023/11/13 09:54:54 [process:4] Cleaning up process
2023/11/13 09:54:54 [process:4] Starting process
2023/11/13 09:54:54 [process:4] Closing process's stdin
2023/11/13 09:54:54 [process:4] Process exited with code 0
2023/11/13 09:54:54 [process:4] Cleaning up process
2023/11/13 09:54:54 [harbor:4] Client logged in.

Then when closing strange things start to happen:
 
2023/11/13 09:55:02 [source:4] Source source.11 gets down.
2023/11/13 09:55:02 [source:4] Source audio.add gets down.
2023/11/13 09:55:02 [source:4] Source fade_in gets down.
2023/11/13 09:55:02 [source:4] Source track_amplify gets down.
2023/11/13 09:55:02 [source:4] Source switch.4 gets down.
2023/11/13 09:55:02 [source:4] Source fail gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:02 [fail:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [blank.strip.4:4] Setting state to blank
2023/11/13 09:55:04 [switch.3:3] Switch to switch.2 with forgetful transition.
2023/11/13 09:55:04 [source:4] Source sequence gets down.
2023/11/13 09:55:04 [source:4] Source max_duration gets down.
2023/11/13 09:55:04 [source:4] Source fail gets down.
2023/11/13 09:55:04 [source:4] Source sequence.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source max_duration.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source source.14 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source.14:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source audio.add.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [audio.add.2:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source fade_in.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [fade_in.2:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source track_amplify.3 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [track_amplify.3:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source replay_metadata.3 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [replay_metadata.3:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source switch.5 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source fade_out.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [fade_out.2:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source track_amplify.4 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:55:04 [track_amplify.4:3] Content type is {audio=pcm(stereo)}.
2023/11/13 09:55:04 [source:4] Source fail.2 gets up with content type: {audio=pcm(stereo)}.
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 885988)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..
2023/11/13 09:54:37 [live:4] Generator max length exeeded (882000 < 886096)! Dropping content..

[Message repeated until script stopped; during generating this message CPU usage of the script keeps on increasing]

What I'm curious about (that's why I've categorized it as a question) is 
1. Is there some logical reason why the "Generator max length exceeded" message has a timestamp that's older as the previous log lines?
2. What does this "Generator max length exceeded" message really mean? In the source I can see that it's metadata related, but can't find some kind of documentation for it

Maybe the answers to these questions help me finding the reason for this to happen - and solving it or being able to file a proper bug report ;-)

**Comment by toots** [comment]
Hi @Calypso971, thanks for reporting. It does look like something is blocking. The max length message would typically be delivered if the live source keeps receiving data but this data is not connected.

I think that, to be able to give more informations, we would need to see more of your script.

**Comment by Calypso971** [comment]
No problem. The log is not from this script below, but I've had the same issue with the one below. The main problem is: it isn't reproducable: it sometimes happens, but most of the time it's not. However, when it happens, the output goes simply blank because it's too CPU heavy. So basically your mention of something blocking is making sense: I have the feeling that it's a more or less rare condition that leads to it.

The script almost can't be more basic - maybe the issue is in the process.run - I can imagine that one is in some way blocking stuff:

def dj_leave() =
  log(level=2, "[DJ LEAVE] Leaving")
  process.run(timeout=2.,"/etc/liquidsoap/scripts/dj_leave.sh &")
end

live = input.harbor("live", port=8300, id="live", on_disconnect=dj_leave, user="source", password="test123", buffer=5., max=20., replay_metadata=true)
security = mksafe(playlist(reload_mode="watch", "/var/local/bla/bla.m3u", id="security")

def on_meta(n,s) =
  def fn(m) =
    print("[GOT METADATA] Got metadata for #{n}: #{m}")
  end
  s.on_metadata(fn)
end

on_meta("live", live)

output.dummy(fallible=false, security)
output.dummy(fallible=true, live)

stream = fallback(replay_metadata=true, track_sensitive=false, [live,security])
on_meta("output", stream)

output.icecast(
            %mp3(bitrate=192, stereo=true, samplerate=44100, internal_quality=2),
            host="127.0.0.1", port = 9000,
            public = false, fallible = true,
            description="Description",
            mount="/feed",
            send_icy_metadata=true,
            icy_metadata=["url","title","artist"],
            password = "mypass",
            url = "https://www.bla.com",
            name = "Great slogan",
            genre = "Pop",
            stream)

**Comment by toots** [comment]
Thanks for this. In a situation like this, I would suspect the `output.icecast` being blocked on some network operation and `input.harbor` getting saturated with data while the output is blocked.

Buffers were rewritten in `2.2.x` so it's possible the problem was already there in `2.1.x` but not logged in properly.

I would suggest to separate the sources with a buffer to place them in different clocks. Also, the dummy output are probably not required anymore. Something like this:
```liquidsoap
def dj_leave() =
  log(level=2, "[DJ LEAVE] Leaving")
  process.run(timeout=2.,"/etc/liquidsoap/scripts/dj_leave.sh &")
end

live = input.harbor("live", port=8300, id="live", on_disconnect=dj_leave, user="source", password="test123", buffer=5., max=20., replay_metadata=true)
security = mksafe(playlist(reload_mode="watch", "/var/local/bla/bla.m3u", id="security")

def on_meta(n,s) =
  def fn(m) =
    print("[GOT METADATA] Got metadata for #{n}: #{m}")
  end
  s.on_metadata(fn)
end

on_meta("live", live)

live = buffer(mksafe(live))

stream = fallback(replay_metadata=true, track_sensitive=false, [live,security])
on_meta("output", stream)

output.icecast(
            %mp3(bitrate=192, stereo=true, samplerate=44100, internal_quality=2),
            host="127.0.0.1", port = 9000,
            public = false, fallible = true,
            description="Description",
            mount="/feed",
            send_icy_metadata=true,
            icy_metadata=["url","title","artist"],
            password = "mypass",
            url = "https://www.bla.com",
            name = "Great slogan",
            genre = "Pop",
            stream)
```

**Comment by Calypso971** [comment]
Thanks for the explanation - will try the different clock scenario to see if it happens again. The script was adapted from a running 2.1 script (changes would only reflect the changes needed for the version change) and never experienced the CPU hogging behaviour. 

I've tried the script without dummy outputs in 2.2 (as a test to see if it was related and following your remark about propably not needed that I've picked up in other issue tickets), but I noticed that Metadata changes in non-used streams weren't picked up then. (Scenario: 2 input streams, stream 1 is active, processes incoming metadata "a", stream 2 becomes active, processes incoming metadata, in the meanwhile input of stream 1 changes metadata to "b", stream 2 disconnects, stream 1 shows metadata "a"). With a dummy output they do.
